File: graceful-florist-web/src/app/modules/products/models/product.dto.ts
import {
  AbstractAuditableDTO,
  AbstractBaseDto,
  SearchResultDto
} from '../../shared/models/abstract-base-dto';
import {CategoryDto, CategoryType} from '../../admin/model/category.dto';
import {uuid} from '../../../../../graceful-florist-type';

export interface ProductDto extends AbstractAuditableDTO {
  name: string;
  price: number;
  purchases: number;
  mainImage: uuid;
  images: uuid[];
  enabled: boolean;
  description: string;
  notes: string;
  owner?: UserDto;
  customPrice?: ProductCustomPriceDto;
}

export interface ProductCustomDTO {
  price: number;
  description: string;
  categories: CategoryDto[];
  ingredients: IngredientDto[];
  customPrice: ProductCustomPriceDto;
  notes: string;
}

export enum ProductStatus {
  SELLING = 'SELLING',
  NOT_SELLING = 'NOT_SELLING'
}

export interface ProductDetailDto extends ProductDto {
  reviewCount: number;
  purchaseCount: number;
  ingredients: IngredientDto[];
  description: any;
  comments: SearchResultDto<CommentDto>;
}

export interface IngredientDto {
  id: uuid;
  name: string;
  image: uuid;
  price: number;
  type: IngredientType;
  quantity?: number;
}

export interface CommentDto {
  id: string;
  content: string;
  rating: number;
  images: string[];
  address: string;
  createdDate: string;
  user: UserDto;
}

export interface CommentSearchCriteriaDto {
  productId: string;
}

export interface UserDto extends AbstractAuditableDTO {
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  avatar: uuid;
}

export interface ProductCriteriaDto {
  keyword?: string;
  status?: ProductStatus;
  fromInclusive?: Date;
  categoryType?: CategoryType;
  categories: CategoryDto[];
}

export enum IngredientType {
  MAIN_FLOWER = 'MAIN_FLOWER',
  SECONDARY_FLOWER = 'SECONDARY_FLOWER',
  BACKGROUND_FLOWER = 'BACKGROUND_FLOWER',
  ACCESSORIES = 'ACCESSORIES',
  LAYOUT = 'LAYOUT',
  PACKAGING = 'PACKAGING'
}

export interface ProductCustomPriceDto extends AbstractBaseDto {
  name: string;
  price: number;
}

File: graceful-florist-services/src/main/java/id/vn/thongdanghoang/graceful/dtos/products/ProductDTO.java
package id.vn.thongdanghoang.graceful.dtos.products;

import id.vn.thongdanghoang.graceful.dtos.AbstractAuditableDTO;
import id.vn.thongdanghoang.graceful.dtos.users.UserDTO;
import lombok.Data;

import java.util.Set;
import java.util.UUID;

@Data
public class ProductDTO extends AbstractAuditableDTO {
    private Set<UUID> images;
    private UUID mainImage;
    private String name;
    private String description;
    private String notes;
    private int price;
    private long quantity;
    private int purchases;
    private int rating;
    private boolean enabled;
    private Set<CategoryDTO> categories;
    private Set<IngredientDTO> ingredients;
    private UserDTO owner;
    private ProductCustomPriceDto customPrice;
}
